engine/src/board.rs

```diff
@@ -200,16 +213,39 @@ impl Board {
         };
     }
 
-    pub fn is_top_piece(&self, piece: &Piece, position: &Position) -> bool {
-        self.board.get(position).unwrap().last().unwrap() == piece
+    // Use this whenever you interact with untrusted input, e.g. user input, frontend input,
+    // history files, etc.
+    pub fn top_piece(&self, position: &Position) -> Result<Piece, GameError> {
```

Comment:
similar to `position()`, i feel like a reasonable return value for this would be `Option<Piece>`, where `None` signifies there's no pieces at that position -- leave it to the caller to decide if that's an error or not